modules/4337/contracts/Safe4337Module.sol

```diff
@@ -69,6 +69,31 @@ contract Safe4337Module is IAccount, HandlerContext, CompatibilityFallbackHandle
         address entryPoint;
     }
 
+    /**
+     * @dev A signed Safe operation, this data is emitted as part of an event when Safe operations get executed.
+     */
+    struct SignedSafeOp {
+        address safe;
+        uint256 nonce;
+        bytes initCode;
+        bytes callData;
+        uint256 callGasLimit;
+        uint256 verificationGasLimit;
+        uint256 preVerificationGas;
+        uint256 maxFeePerGas;
+        uint256 maxPriorityFeePerGas;
+        bytes paymasterAndData;
+        uint48 validAfter;
+        uint48 validUntil;
+        address entryPoint;
+        bytes signatures;
+    }
+
+    /**
+     * @notice An event that gets emitted when a Safe operation executes.
+     */
+    event SafeOperation(SignedSafeOp operation);
```

Comment:
Nothing needs to be indexed in this event, right?

Reply:
That is a question for @Uxio0

Reply:
I think indexing the safe address could be important for the future (I'm thinking for example in Dune analytics). Also would it be posible to emit the `SafeOpHash` or is it too much?

Reply:
I think indexing the safe address could be important for the future (I'm thinking for example in Dune analytics). Also would it be posible to emit the `SafeOpHash` or is it too much?

Reply:
As long as we trust the entrypoint event not to change and as we can correlate entrypoint events with ours (I don't know if it can be done), it's alright

Reply:
The point here is not what information we have on the entrypoint. The issue is that entrypoint is missing crucial data (the signature) to build the Safe operation, and as we already have to emit an event we add all the information to build the SafeOperation without depending on the entrypoint. All the data but the signature can be obtained from the entrypoint, so following that logic we should only emit an event with the signature

Reply:
One possible alternative is to emit:
1. UserOpHash (so that it can be indexed against a user operation [event](https://github.com/eth-infinitism/account-abstraction/blob/fa61290d37d079e928d92d53a122efcc63822214/contracts/interfaces/IEntryPoint.sol#L29))
2. Signature bytes.

The issue I see, however, is that even user operation data is not emitted as part of events, so some additional indexing (like parsing calldata) might still be necessary. Additionally, AFAIU, the entry point is not guaranteed to be called by an EOA, so it might even require tracing in order to get the calldata to extract the user operation parameters.

Reply:
I'm ok with only indexing the missing data as long as you emit `UserOpHash` so I can match the information in the entrypoint, but we cannot count on relying on tracing, as it's not available for some L2 networks

Reply:
If we're only missing a signature from the event emitted by the entrypoint, I'd contact the 4337 team to see if they can consider adding it. They also have a release coming, so I'd hurry up with this suggestion.

Reply:
We are missing also the `initCode`, `callData`, `data` for the paymaster... If they are willing to add all the data to the event that would be awesome, but I don't think they would be willing to

Reply:
Hmm, somehow https://jiffyscan.xyz/userOpHash/0xd0c9c8761cd347d0642741c7998be32ab2855471dce4a3c28727fa0ba49dfc4c?network=mainnet can index it. Perhaps we can reach out to them to learn how they implemented it

Reply:
I can do that for mainnet too, using tracing 😉

Reply:
![Screenshot 2024-02-11 at 17 47 34](https://github.com/safe-global/safe-modules/assets/16622558/2e9fbc9a-ad5a-4418-b390-5f8def2bff01)

Here's the list of networks that they support - guess that's doable without tracing

Reply:
I guess then they are using the bundler APIs https://docs.alchemy.com/reference/bundler-api-endpoints